### PR TITLE
Override keyword throughout Framework

### PR DESCRIPTION
**Pull request by permcody**
The primary purpose of this PR is to add the keyword override throughout the framework. The "base" directory is not included in this set because it is very large an also causes a few compile errors with the `Outputs` system which I am still looking at. Additionally, I have fixed several minor issues with access protection throughout and removed all unneeded virtual destructors in derived classes.

refs #7171

### CODE REVIEW

**Comment by permcody** [comment]
@idaholab/moose-team - This PR will have to be merged fairly soon since it touches nearly everything. If you don't like it, let me know soon!

**Comment by friedmud** [comment]
So - from this point forward will leaving out `override` cause an error?  Because these are just going to creep back in...

**Comment by dschwen** [comment]
:+1: for  getting rid of unnecessary empty virtual destructors and fixing access qualifiers! override is also nice. This PR is too big to fully review, but at a first glance it looks fine.

File: framework/include/multiapps/TransientMultiApp.h [hunk 1]
```diff
@@ -36,53 +36,23 @@ class TransientMultiApp :
 
   virtual ~TransientMultiApp();
 
-  /**
```
Comment (YaqiWang) [inline framework/include/multiapps/TransientMultiApp.h]: is this comment to be removed?

File: framework/include/multiapps/TransientMultiApp.h [hunk 2]
```diff
@@ -36,53 +36,23 @@ class TransientMultiApp :
...
+  virtual bool needsRestoration() override;
+
+  virtual void resetApp(unsigned int global_app, Real time) override;
 
   /**
    * Finds the smallest dt from among any of the apps.
    */
   Real computeDT();
 
-  /**
```
Comment (YaqiWang) [inline framework/include/multiapps/TransientMultiApp.h]: same here?

**Comment by jwpeterson** [comment]
> So - from this point forward will leaving out override cause an error? Because these are just going to creep back in...

Not errors, but clang does warn about it I believe.

File: framework/include/multiapps/TransientMultiApp.h (hunk 1 above)
Comment (jwpeterson) [inline framework/include/multiapps/TransientMultiApp.h]: @permcody removed comments from newly-overrided functions that were copied/pasted from base class functions.  The idea is to reduce the amount of documentation upkeep overhead, i.e. if you change the base class function and update it's documentation, you don't want to have to remember to go update all the child class documentation.

Comment (YaqiWang) [inline framework/include/multiapps/TransientMultiApp.h]: oh, i see.

**Comment by permcody** [comment]
`override` acts per class (not per complication unit). If you add it in one place, clang will warn about all other overridden functions missing that keyword.

**Comment by permcody** [comment]
This failed ferret but I put in a PR to fix that issue. Should be ready to merge.

**Pull request merged**